### PR TITLE
Enable show more files in diff for git <2.31

### LINKED ISSUE #17731: Internal server error when clicking "Show More" on large PR diff
### Gitea Version

1.16.0+dev-558-ga8fd76557

### Git Version

2.18.0

### Operating System

Windows

### How are you running Gitea?

Built myself.

### Database

_No response_

### Can you reproduce the bug on the Gitea demo site?

Yes

### Log Gist

https://gist.github.com/jpraet/3803f61d48e8db2ee12cfaa354a91135

### Description

https://try.gitea.io/jpraet-org/test-repo/pulls/1/files -> click "Show More".

In network tab of browser console: 
XHR GET https://try.gitea.io/jpraet-org/test-repo/pulls/1/files?skip-to=101.txt&file-only=true
[HTTP/2 500 Internal Server Error 190ms]

### Screenshots

_No response_

### PR DESCRIPTION
**Pull request by zeripath**
Unfortunately due to a misread on my behalf I missed that git diff only learned
--skip-to in version 2.31.0. Thus this functionality was not working on older versions
of git.

This PR adds a handler that simply allows for us to skip reading the diffs until
we find the correct file to skip to.

Fix #17731

Signed-off-by: Andrew Thornton <art27@cantab.net>

### CODE REVIEW

File: services/gitdiff/gitdiff.go [hunk 1]
```diff
@@ -739,6 +741,20 @@ parsingLoop:
 		}
 
 		curFile = createDiffFile(diff, line)
+		if skipping {
```
Comment (Gusted) [inline services/gitdiff/gitdiff.go:742]: Does line 729 still makes sense now?

File: services/gitdiff/gitdiff.go [hunk 2]
```diff
@@ -739,6 +741,20 @@ parsingLoop:
...
+			if curFile.Name != skipTo {
```
Comment (Gusted) [inline services/gitdiff/gitdiff.go]: If `--skip-to` was used, this should technically never evaluate to true?

File: services/gitdiff/gitdiff.go [hunk 3]
```diff
@@ -955,6 +971,36 @@ parsingLoop:
...
+	for {
+		lineBytes, isFragment, err = input.ReadLine()
+		if err != nil {
+			return
+		}
+		if wasFragment {
+			wasFragment = isFragment
+			continue
+		}
+		if bytes.HasPrefix(lineBytes, []byte(cmdDiffHead)) {
```
Comment (Gusted) [inline services/gitdiff/gitdiff.go:985]: `[]byte(cmdDiffHead)` should this be a constant?

File: services/gitdiff/gitdiff.go [hunk 4]
```diff
@@ -694,9 +694,11 @@ func (diff *Diff) LoadComments(issue *models.Issue, currentUser *models.User) er
 const cmdDiffHead = "diff --git "
 
 // ParsePatch builds a Diff object from a io.Reader and some parameters.
-func ParsePatch(maxLines, maxLineCharacters, maxFiles int, reader io.Reader) (*Diff, error) {
+func ParsePatch(maxLines, maxLineCharacters, maxFiles int, reader io.Reader, skipTo string) (*Diff, error) {
```
Comment (Gusted) [inline services/gitdiff/gitdiff.go]:
```suggestion
func ParsePatch(maxLines, maxLineCharacters, maxFiles int, reader io.Reader, skipToFile string) (*Diff, error) {
```

File: services/gitdiff/gitdiff.go (hunk 3 above)
Comment (zeripath) [reply services/gitdiff/gitdiff.go:985]: Go will compile that away.

File: services/gitdiff/gitdiff.go (hunk 2 above)
Comment (zeripath) [reply services/gitdiff/gitdiff.go]: It doesn't matter. I can change it so that it skip-to is used skipping is set to false but the code works as is.

**Review by wxiaoguang — APPROVED** [review]

**Review by Gusted — APPROVED** [review]

**Comment by jpraet** [comment]
Validated with Git version 2.18.

**Pull request merged**